CLI: Fix throwing in readonly environments

JReinhold opened the pull request:
Closes #31709

<!-- If your PR is related to an issue, provide the number(s) above; if it resolves multiple issues, be sure to break them up (e.g. "closes #1000, closes #1001"). -->

<!--

Thank you for contributing to Storybook! Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `main` branch as part of the release process, so you shouldn't need to worry about this. For additional guidance: https://storybook.js.org/docs/contribute

-->

## What I did

1. Changed the global settings logic to only log a warning if it can't save the file, instead of throwing an error
2. Don't run global settings logic in CI
3. Added a utility to more resiliently check for optional, boolean environment variables, and a shortcut `isCI` that uses that.

<!-- Briefly describe what your PR does -->

## Checklist for Contributors

### Testing

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to communicate how to test your changes -->

#### The changes in this PR are covered in the following automated tests:

- [ ] stories
- [x] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

<!-- Please include the steps to test your changes here. For example:

1. Run a sandbox for template, e.g. `yarn task --task sandbox --start-from auto --template react-vite/default-ts`
2. Open Storybook in your browser
3. Access X story

-->

### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli-storybook/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

  - `bug`: Internal changes that fixes incorrect behavior.
  - `maintenance`: User-facing maintenance tasks.
  - `dependencies`: Upgrading (sometimes downgrading) dependencies.
  - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
  - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
  - `documentation`: Documentation **only** changes. Will not show up in release changelog.
  - `feature request`: Introducing a new feature.
  - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
  - `other`: Changes that don't fit in the above categories.

   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

<!-- CANARY_RELEASE_SECTION -->

<!-- BENCHMARK_SECTION -->
<!-- BENCHMARK_SECTION -->


<!-- greptile_comment -->

## Greptile Summary

Improves error handling in read-only environments by replacing error throwing with warning logs and adding robust environment variable handling utilities.

- Added new utilities `isCI` and `optionalEnvToBoolean` in `code/core/src/common/utils/envs.ts` for consistent environment detection
- Modified `code/core/src/telemetry/storybook-metadata.ts` to skip global settings in CI environments
- Updated `code/lib/cli-storybook/src/bin/index.ts` to use warning logs instead of throwing errors when saving fails
- Fixed boolean logic in sandbox initiation to correctly disable dev mode in CI environments



<!-- /greptile_comment -->

JReinhold committed:
dont throw on global settings write error, dont run in CI

shilman reviewed (approved):
Looks good with one minor nit

JReinhold commented on code/core/src/telemetry/storybook-metadata.ts:
```diff
@@ -57,10 +57,10 @@ export const computeStorybookMetadata = async ({
   configDir: string;
 }): Promise<StorybookMetadata> => {
-  const settings = await globalSettings();
+  const settings = process.env.CI === 'true' ? undefined : await globalSettings();
```
`Boolean('false')` returns `true`, seems pretty bad to me.

Ie. `CI=false storybook dev` would make that thing become `true`.

Although `process.env.CI === 'true'` wouldn't support `CI=1`, but I'm not sure how many use that.

tmeasday replied:
Should we use this https://github.com/storybookjs/storybook/blob/c2d246ce49e181fef17a86b2248f3ae538e5fb70/code/core/src/core-server/presets/common-preset.ts#L173-L188

JReinhold committed:
add more resilient utility for boolean environment variables

tmeasday reviewed (approved):
I have a couple questions but looks good otherwise, good improvement.

JReinhold commented on code/core/src/common/utils/envs.ts:
```diff
@@ -67,3 +67,24 @@ export const stringifyProcessEnvs = (raw: Record<string, string>): Record<string
+ * Doing Boolean(process.env.CI) or !process.env.CI is not enough, because users might set CI=false
+ * or CI=0, which would be truthy, and thus return true in those cases.
+ */
+export const isCI = optionalEnvToBoolean(process.env.CI);
```
No the `storybook/internal/common` entrypoint is only usable in Node, not in the browser.

diagramatics commented:
👋 Just wanna bring this back since it was automatically closed. We do need this fix for our CI use case same as #31709 and it would be nice to have this also released to v8 as well since `globalSettings` was introduced in `v8.6.14`.

valentinpalkovic commented:
@JReinhold Can you take another look and maybe resolve the conflicts?

shilman committed:
Merge branch 'next' into jeppe/fix-global-settings-readonly

yannbf committed:
Merge branch 'next' into jeppe/fix-global-settings-readonly

yannbf committed:
Refactor: Update isCI usage to function calls across multiple files

- Changed instances of `isCI` to `isCI()` in various files to ensure consistent function usage.
- Updated related logic in `postinstall-logger.ts`, `postinstall.ts`, `withTelemetry.ts`, and others to reflect this change.
- Enhanced test coverage for `isCI` behavior in `storybook-metadata.test.ts`.

yannbf committed:
Merge branch 'next' into jeppe/fix-global-settings-readonly

yannbf committed:
Comment out userSince test in event-log-checker for discussion on CI value provision

yannbf committed:
bundle in utilities

yannbf committed:
Refactor: Update imports in initiate.ts to use centralized utilities

- Replaced direct imports of `isCI` and `logger` from internal paths with imports from centralized utility modules.
- This change enhances maintainability and consistency across the codebase.

shilman committed:
Fix test install in agentic environments

yannbf committed:
lazy load lazy-universal-dotenv
